### PR TITLE
Docstrings

### PR DESCRIPTION
**Pull request by ianspektor**
- [X] Write docstring guidelines
- [X] Write example docstring (core lag & leak operators)
- [x] Write all remaining docstrings for public API
- [X] Do some work towards improving existing docstrings in internal classes and functions

Tried to strike some balance between not being too opinionated but not leaving too much space for creativity either, since it's consistency we're after 🤓 eager to get feedback

### CODE REVIEW

File: docs/docstring_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    Leave one blank line.  The rest of this docstring should contain an
+    overall description of the module.  Optionally, it may also contain a brief
+    description of exported classes and functions and/or usage examples.
+
+    Typical usage example:
+
+    foo = ClassFoo()
+    bar = foo.FunctionBar()
+    """
```
Comment (achoum) [inline docs/docstring_guidelines.md:84]: Do we allow md formatting in docstring? For example!:

"""
A one-line summary of the module, terminated by a period.

  Leave one blank line.  The rest of this docstring should contain an
  overall description of the module.  Optionally, it may also contain a brief
  description of exported classes and functions and/or usage examples.
  Typical usage example:

```python
  foo = ClassFoo()
  bar = foo.FunctionBar()
```

"""

File: temporian/core/data/event.py [hunk 2]
```diff
@@ -22,6 +22,27 @@
 
 
 class Event(object):
+    """Collection of feature values for a certain sampling.
+
+    An event can be thought of as a list of multivariate series, where each item
```
Comment (achoum) [inline temporian/core/data/event.py:27]: s/list/dictionary?

File: temporian/core/data/event.py [hunk 3]
```diff
@@ -22,6 +22,27 @@
 
 
 class Event(object):
```
Comment (achoum) [inline temporian/core/data/event.py:24]: Nice text. I tried some improvements (with the help of chatgpt) :):

An *event* represents the structure, or schema, of a collection of indexed multivariate time series. An event does not contain any actual data, but is instead used as a reference to describe the format of the input, intermediate results, or output of a *Processor* (i.e., a computation graph).

Informally, an event defines the name and data types of each time series, as well as the key and data type of the index (if any).

There are several ways to create an event:

- When importing data from a Pandas DataFrame with the `a = NumpyEvent.from_dataframe(...)` method, the event is automatically constructed and available with the `a.schema()` method.
- Manually using the `temporian.input_event(...)` method to specify the name and data types of each time series and the key and data type of the index.
- (Not recommended) By instantiating the Event class directly.

File: temporian/core/evaluator.py [hunk 4]
```diff
@@ -38,26 +38,26 @@ def evaluate(
...
 
     Args:
-        query: Events to compute. Support event, dict and list of events.
+        query: Events to compute. Supports event, dict and list of events.
         input_data: Dictionary of event and event values to use for the
-          computation.
-        verbose: If >0, prints details about the execution on the standard error
-          output. The larger the number, the more information is displayed.
+            computation.
+        verbose: If > 0 prints details about the execution on the standard error
```
Comment (achoum) [inline temporian/core/evaluator.py:47]: comma?

**Review by achoum — APPROVED** [review]
Awesome!

File: docs/docstring_guidelines.md (hunk 1 above)
Comment (ianspektor) [reply docs/docstring_guidelines.md:84]: not sure if mkdocs supports it - cc @MNLucas

File: temporian/core/serialize.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def save(
     outputs: processor.MultipleEventArg,
     path: str,
 ) -> None:
-    """Saves the computation between "inputs" and "outputs" into a file.
+    """Saves the processor between `inputs` and `outputs` to a file.
```
Comment (getchebarne) [inline temporian/core/serialize.py:36]: suggestion: reword to "containing a set of `inputs`, `outputs`, and `operators`" instead of "between `inputs` and `outputs`", as the `Processor` object contains all of these objects

File: temporian/implementation/numpy/evaluator.py [hunk 6]
```diff
@@ -29,6 +29,17 @@ def evaluate_schedule(
     verbose: int,
     check_execution: bool,
 ) -> Dict[event.Event, numpy_event.NumpyEvent]:
+    """Evaluates a schedule on a dictionary of input event data.
+
+    Args:
+        inputs: Mapping of core Events to materialized NumpyEvents.
+        schedule: sequence of operators to apply on the data.
```
Comment (getchebarne) [inline temporian/implementation/numpy/evaluator.py:36]: capitalize `sequence`

File: temporian/implementation/numpy/evaluator.py [hunk 7]
```diff
@@ -29,6 +29,17 @@ def evaluate_schedule(
...
+        verbose: If >0, prints details about the execution on the standard error
+            output. The larger the number, the more information is displayed.
+        check_execution: If True, data of the intermediate results of the
+            operators is checked against its expected structure and raises if
+            it difers.
```
Comment (getchebarne) [inline temporian/implementation/numpy/evaluator.py:41]: typo `difers` -> `differs`

File: temporian/implementation/numpy/evaluator.py [hunk 8]
```diff
@@ -29,6 +29,17 @@ def evaluate_schedule(
...
 ) -> Dict[event.Event, numpy_event.NumpyEvent]:
+    """Evaluates a schedule on a dictionary of input event data.
+
+    Args:
+        inputs: Mapping of core Events to materialized NumpyEvents.
+        schedule: sequence of operators to apply on the data.
+        verbose: If >0, prints details about the execution on the standard error
+            output. The larger the number, the more information is displayed.
+        check_execution: If True, data of the intermediate results of the
+            operators is checked against its expected structure and raises if
```
Comment (getchebarne) [inline temporian/implementation/numpy/evaluator.py:40]: `and raises` -> `raising an error`

File: temporian/implementation/numpy/operators/arithmetic.py [hunk 9]
```diff
@@ -24,20 +24,23 @@
...
 
         Args:
-            event_1: First Event.
-            event_2: Second Event.
+            event_1: First event.
+            event_2: Second event.
 
         Returns:
-            Arithmetic of the two Events according to resolution and arithmetic operator.
+            Result of the arithmetic operation of the two Events according to
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/arithmetic.py:42]: nit: lowercase "Events" for consistency with arguments?

File: temporian/implementation/numpy/operators/calendar/base.py [hunk 10]
```diff
@@ -53,16 +53,17 @@ def __call__(self, sampling: NumpyEvent) -> Dict[str, NumpyEvent]:
...
-        specific calendar operator. E.g., calendar_day_of_month will take the
-        datetime's day, and calendar_hour will take the its hour.
+        """Gets the value of the datetime object that corresponds to each
+        specific calendar operator.
 
-        Must be implemented by subclasses.
+        For example, calendar_day_of_month will take the datetime's day, and
+        calendar_hour will take the its hour.
+
+        Returned value is be converted to int32 by __call__.
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/calendar/base.py:62]: remove `be`

File: temporian/implementation/numpy/operators/calendar/base.py [hunk 11]
```diff
@@ -53,16 +53,17 @@ def __call__(self, sampling: NumpyEvent) -> Dict[str, NumpyEvent]:
...
     def _get_value_from_datetime(self, dt: datetime) -> Any:
-        """Get the value of the datetime object that corresponds to each
-        specific calendar operator. E.g., calendar_day_of_month will take the
-        datetime's day, and calendar_hour will take the its hour.
+        """Gets the value of the datetime object that corresponds to each
+        specific calendar operator.
 
-        Must be implemented by subclasses.
+        For example, calendar_day_of_month will take the datetime's day, and
+        calendar_hour will take the its hour.
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/calendar/base.py:60]: nit: `take` -> `get` for consistency

File: temporian/implementation/numpy/operators/window/base.py [hunk 12]
```diff
@@ -76,10 +76,10 @@ def _build_accumulator_mask(
     ) -> np.array:
         """Creates a 2d boolean matrix containing the summing instructions.
 
-        The returned matrix "m[i,j]" is defined as:
-            m[i,j] is true iif. input value "j" is averaged in the output value "i".
+        Returns:
+            Matrix "m[i,j]" defined as: m[i,j] is true iif. input value "j" is
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/window/base.py:80]: i'd reword to ```"Matrix `m`, where `m[i,j]` is true iif ..."```

File: temporian/implementation/numpy/operators/window/base.py [hunk 13]
```diff
@@ -111,23 +111,17 @@ def _apply_accumulator_mask(
...
-            values: A 2D NumPy array with shape (n, m) where each row represents
-                a  window of data points. 'n' is the length of the feature, and
-                'm' is the size of the window. The array can contain 'nan'
-                values as padding.
+            values: the input array, of shape (n, m).
 
         Returns:
-            np.array: A 1D NumPy array with shape (n,) containing the operation
-                    for each row (window) in the input array.
+            A 1D NumPy array of shape (n,) with the result for each row in the
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/window/base.py:125]: `with` -> `containing`, `for` -> `of`

**Review by getchebarne — APPROVED** [review]
Great job 💯💯  one more comment:

in some cases, we use a "The" preposition when describing function arguments in a docstring, e.g.
```
window_length: The length of the window
```
I'd remove the "The" and just leave the "Length of the window" part - makes it quicker to read. We already do this in some cases but these are the minority.

File: temporian/core/data/feature.py [hunk 14]
```diff
@@ -21,6 +21,23 @@
 
 
 class Feature(object):
+    """Feature values for a certain sampling.
+
+    A feature can be thought of as a list of univariate series, where each item
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/data/feature.py:26]: ignoring since Feature class no longer exists

File: temporian/core/operators/calendar/hour.py [hunk 15]
```diff
@@ -39,15 +35,16 @@ def output_feature_name(cls) -> str:
 
 
 def calendar_hour(sampling: Event) -> Event:
-    """Obtain the hour each of the timestamps in an event's sampling belongs to.
-    Features in input event are ignored. Output feature contains numbers
-    between 0 and 23.
+    """Obtains the hour the timestamps in an event's sampling are in.
+
+    Features in input event are ignored. Output feature contains numbers between
+    0 and 23.
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/operators/calendar/hour.py:41]: because that's what python datetime's class does - https://docs.python.org/3/library/datetime.html

File: temporian/core/operators/sample.py [hunk 16]
```diff
@@ -99,10 +95,10 @@ def sample(
 
     Args:
         event: The event to sample.
-        sampling: The timestamp of the sampling.
+        sampling: The event to use the sampling of.
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/operators/sample.py:98]: the `sampling` argument is indeed an `Event`

File: temporian/core/operators/window/moving_standard_deviation.py [hunk 17]
```diff
@@ -59,30 +46,30 @@ def moving_standard_deviation(
     window_length: Duration,
     sampling: Optional[Event] = None,
 ) -> Event:
-    """Moving Standard Deviation
+    """Computes the standard deviation of values in a sliding window over the
+    event.
 
     For each t in sampling, and for each feature independently, returns at time
     t the standard deviation for the feature in the window
     [t - window_length, t].
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/operators/window/moving_standard_deviation.py:54]: by looking at the code, it would fail when doing np.nanstd() in numpy impl. not addressing in this PR to limit scope.
